File: Stack/EvaluateReversePolishNotation/erpn.py
from typing import List


class Solution:
    # NOTE: Time = O(n), Space = O(n)
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        oprts = {"+", "-", "*", "/"}

        for t in tokens:
            if t in oprts:
                b = stack.pop()
                a = stack.pop()
                res = self._calc(a, b, t)
                stack.append(res)
            else:
                stack.append(int(t))
            print(stack)

        return stack.pop()

    def _calc(self, a: int, b: int, op: str) -> int:
        if op == "+":
            return a + b
        elif op == "-":
            return a - b
        elif op == "*":
            return a * b
        else:
            return int(a / b)


if __name__ == "__main__":
    print("MAIN:")
    sol = Solution()

    # ANS: 5
    tks = ["1", "2", "+", "3", "*", "4", "-"]
    r = sol.evalRPN(tks)
    print(f"T1: {r}")

    # ANS: 22
    tks = ["10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"]
    r = sol.evalRPN(tks)
    print(f"T2: {r}")

File: Stack/EvaluateReversePolishNotation/erpn.go
package main

import (
	"fmt"
	"strconv"
)

func evalRPN(tokens []string) int {
	stack := []int{}
	for _, t := range tokens {
		switch t {
		case "+", "-", "*", "/":
			b := stack[len(stack)-1]
			stack = stack[:len(stack)-1]

			a := stack[len(stack)-1]
			stack = stack[:len(stack)-1]

			switch t {
			case "+":
				stack = append(stack, a+b)
			case "-":
				stack = append(stack, a-b)
			case "*":
				stack = append(stack, a*b)
			case "/":
				stack = append(stack, a/b)
			}
		default:
			n, _ := strconv.Atoi(t)
			stack = append(stack, n)
		}
	}
	return stack[len(stack)-1]
}

func main() {
	// ANS: 5
	tks := []string{"1", "2", "+", "3", "*", "4", "-"}
	fmt.Printf("T1: %d\n", evalRPN(tks))

	// ANS: 22
	tks = []string{"10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"}
	fmt.Printf("T2: %d\n", evalRPN(tks))
}
